onnx/funconnx/op_run.py

```diff
@@ -0,0 +1,508 @@
+# SPDX-License-Identifier: Apache-2.0
+# pylint: disable=C0415,R0912
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterable, List, Optional, Tuple
+
+import numpy as np  # type: ignore
+
+from ..defs import get_all_schemas_with_history, get_schema, onnx_opset_version
+from ..helper import make_node
+from ..numpy_helper import to_array
+from ..onnx_pb import AttributeProto, GraphProto, NodeProto, TypeProto
+
+
+class RuntimeTypeError(RuntimeError):
+    """
+    Raised when a type of a variable is unexpected.
+    """
+
+
+class DefaultNone:
+    """
+    Default value for parameters when the parameter is not set
+    but the operator has a default behavior for it.
+    """
+
+
+class RefAttrName:
+    """
+    Implements a link between a parameter of a function
+    and an attribute in node.
+
+    :param name: name of the input
+    """
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def __repr__(self) -> str:
+        "usual"
+        return f"{self.__class__.__name__}({self.name!r})"
+
+
+def _build_schemas() -> Dict[str, type]:
+    res: Dict[str, type] = {}
+    for schema in get_all_schemas_with_history():
+        # Multiple version can coexist. The last one is kept.
+        if schema.name in res:
+            if schema.since_version > res[schema.name].since_version:  # type: ignore
+                # We keep the most recent one.
+                res[schema.name] = schema  # type: ignore
+        else:
+            res[schema.name] = schema  # type: ignore
+        res[schema.name + "_" + str(schema.since_version)] = schema  # type: ignore
+    return res
+
+
+_schemas = _build_schemas()
+
+
+class OnnxType:
+    def __init__(self, type_proto: TypeProto):
+        if not isinstance(type_proto, TypeProto):
+            raise TypeError(f"type_proto {type(type_proto)} must be of type TypeProto.")
+        self.type_proto = type_proto
+
+    def __repr__(self) -> str:
+        return f"OnnxType({self.type_proto!r})"
+
+
+class SparseTensor:
+    """
+    Simple representation of a sparse tensor.
+    It is based on numpy but does not require scipy.
+    """
+
+    def __init__(
+        self, values: np.ndarray, indices: np.ndarray, shape: Tuple[int]
+    ) -> None:
+        self.values = values
+        self.indices = indices
+        self.shape = shape
+
+    @property
+    def dtype(self) -> Any:
+        return self.values.dtype
+
+
+def to_sparse_tensor(att: AttributeProto) -> SparseTensor:
+    """
+    Hosts a sparse tensor.
+    """
+    shape = tuple(d for d in att.dims)
+    return SparseTensor(to_array(att.values), to_array(att.indices), shape)  # type: ignore
+
+
+class Graph:
+    def __init__(self, g: GraphProto) -> None:
+        self.g = g
+
+
+class OpRun(ABC):
+    """
+    Ancestor to all operators in this subfolder.
+    The runtime for every node can checked into
+    `ONNX unit tests
+    <https://github.com/onnx/onnx/tree/master/onnx/backend/test/case/node>`_.
```

Comment:
Not sure I understand what the above comment means.

Reply:
I see some duplication of code across the above folder and the op implementations in this PR, which I guess is related. But not sure what the plan/suggestion above is.

Reply:
I updated the comment. I forgot to change it.